### ISSUE TITLE
Plymouth support 

### ISSUE THREAD

**Issue by advaithm** [anatol/booster#44]
Plymouth is a project from Fedora and now listed among the freedesktop.org's official resources providing a flicker-free graphical boot process. It relies on kernel mode setting (KMS) to set the native resolution of the display as early as possible, then provides an eye-candy splash screen leading all the way up to the login manager.

Plymouth is used by many Linux distributions. These include Ubuntu, fedora and pop os to name a few. It would be very useful for down stream distro maintainers to be able to use booster along with plymouth.

**Comment by anatol** [comment]
I was trying to read about how splash screens are implemented and see that there are other ways to do it e.g. `fbsplash`. Is `fbsplash` actively used? What is preferrred plymouth or fbsplash?

**Comment by advaithm** [comment]
Plymouth is used by most mainstream linux distributions. I haven't really heard of  fbsplash so I don't know if any linux distros use it

**Comment by oech3** [comment]
I wonder flicker-free boot is possible with `fbsplash`.

**Comment by emansom** [comment]
I would like to have a flicker free fancy looking boot experience when using Booster. Please implement this!

**Comment by filipkemuel** [comment]
Plymouth has active development..
fbsplash hasn't been updated since 2014..

**Comment by nmeum** [comment]
> I was trying to read about how splash screens are implemented and see that there are other ways to do it e.g. `fbsplash`. Is `fbsplash` actively used? What is preferrred plymouth or fbsplash?

Alpine actually (optionally) uses fbsplash, the [BusyBox version](https://git.busybox.net/busybox/tree/miscutils/fbsplash.c) though: https://gitlab.alpinelinux.org/alpine/mkinitfs/-/blob/9623bc1df5c1e3b861c50d67a4c837b76a20ab85/initramfs-init.in#L670-L697 but plymouth is surely the more popular solution.

**Comment by aleksasiriski** [comment]
+1 for plymouth, I don't want to switch from booster but I would still like to use plymouth (:

**Comment by codefaux** [comment]
Pitching my vote - currently building a handheld distro around Arch and almost got to use Booster for it. Shame.

**Comment by ReFleXzZ** [comment]
+1 for this one. Still nothing? :C

**Comment by nhubaotruong** [comment]
+1 for this, hope that it gets implement soon 🙏

**Comment by ToasterUwU** [comment]
I use PikaOS (im also Part of the team) and we switched to Booster for the faster boot times, but had to sacrifice our custom refind theme for it sadly.

Would be awesome to get the best of both worlds again.

**Comment by Tracker-Friendly** [comment]
evolutionOS developer here: same as above! we are working around this by using the busybox fork of fbsplash, but it's not ideal. plymouth soon? :heart:

**Comment by Tracker-Friendly** [comment]
Make a boot service that runs fbsplash hand an image

**Comment by number3d** [comment]
still no love for plymouth?

**Comment by Arbel-arad** [comment]
doesn't seem like anyone is working on it (or booster updates in general)

**Comment by ferrreo** [comment]
I have a branch with plymouth working somewhat but sadly I am struggling to get it working with LUKs. I can only get plymouth to default back to text mode even when using efi framebuffer/simpledrm. I have tried forcing various modules to be loaded at various times and don't seem to be able to work out the magic combination.

@anatol any advice on working this out? We essentially need simpledrm loaded and framebuffer/screen active but drives not mounted yet when we call to init plymouthd then show splash via plymouth.

I'm also working on some early/late hooks (for ISO boots) that might achieve a similar result without plymouth specific code needed.

**Comment by LDprg** [comment]
@anatol is there any chance we get this implemented? 
While fbsplash works, most distros use Plymouth. It is also kind of annoying if I have a working Plymouth with dracut and want to switch to booster (since I need to uninstall Plymouth and install fbsplash). Another thing is that fbsplash is not updated anymore (which is not inherently bad, but it is not a good thing for sure). Plymouth is also much more customizable.

PS: Great project! I love booster, especially when playing around with a lot of different kernels.

**Comment by Arzumify** [comment]
I've been taking a look at this issue again and I've found out how plymouth _actutally_ works.

> Somehow you need to hook into your distros mkinitrd mechanism.   I
> mean you need to copy the plymouth binaries, plugins, and data files
> into the initrd everytime you rebuild your initrd. See
> plymouth-populate-initrd for how we do it on fedora (mkinitrd calls
> plymouth-populate-initrd).  I would like to upstream hooks for other
> distros mkinitrd systems, so the out of the box experience is easier.

> Inside the initrd, you need to start plymouthd early.  The earlier the
> better, but not before /sys, /proc/,  and /dev/pts are mounted, and
> not before /dev/tty[01] and /dev/fb0 are created.

> At some point during the initrd you need to load the kms driver.
> After it's loaded the graphics mode should get set up and /dev/fb0
> will be functional.  That's when you call plymouth --show-splash to
> show the splash screen.

This doesn't exactly seem like rocket science, for plymouth support all you'd have to do is to load KMS drivers then to run the plymouth binary. Now I'm not very experienced in booster's code, but if I get desperate enough I may try to make a PR.

**Comment by ferrreo** [comment]
> I've been taking a look at this issue again and I've found out how plymouth _actutally_ works.
> 
> > Somehow you need to hook into your distros mkinitrd mechanism.   I
> > mean you need to copy the plymouth binaries, plugins, and data files
> > into the initrd everytime you rebuild your initrd. See
> > plymouth-populate-initrd for how we do it on fedora (mkinitrd calls
> > plymouth-populate-initrd).  I would like to upstream hooks for other
> > distros mkinitrd systems, so the out of the box experience is easier.
> 
> > Inside the initrd, you need to start plymouthd early.  The earlier the
> > better, but not before /sys, /proc/,  and /dev/pts are mounted, and
> > not before /dev/tty[01] and /dev/fb0 are created.
> 
> > At some point during the initrd you need to load the kms driver.
> > After it's loaded the graphics mode should get set up and /dev/fb0
> > will be functional.  That's when you call plymouth --show-splash to
> > show the splash screen.
> 
> This doesn't exactly seem like rocket science, for plymouth support all you'd have to do is to load KMS drivers then to run the plymouth binary. Now I'm not very experienced in booster's code, but if I get desperate enough I may try to make a PR.

Yeah that's what my branch here does: https://github.com/anatol/booster/compare/master...ferrreo:booster:plymouth

Issue I am having is I can't get the gfx mode/kms/fb stuff to work. It always fallback to text mode.

**Comment by pilotstew** [comment]
I spent the last couple days working on this with help from claude. Plymouth is working well for me. For your consideration: https://github.com/pilotstew/booster/tree/plymouth

**Comment by anatol** [comment]
> I spent the last couple days working on this with help from claude. Plymouth is working well for me. For your consideration: https://github.com/pilotstew/booster/tree/plymouth

On the first glance things look good. Thank you @pilotstew.

Could you please send a PR for it?
